src/ol/render/canvas/Executor.js

```diff
@@ -981,28 +915,39 @@ class Executor {
                     ((0.5 - baseline) * 2 * strokeWidth * textScale[1]) /
                       textScale[0] -
                     offsetY;
-                  rendered =
-                    this.replayImageOrLabel_(
-                      context,
-                      contextScale,
-                      /** @type {number} */ (part[0]),
-                      /** @type {number} */ (part[1]),
-                      label,
-                      anchorX,
-                      anchorY,
-                      declutterGroup,
-                      label.height,
-                      1,
-                      0,
-                      0,
-                      /** @type {number} */ (part[3]),
-                      pixelRatioScale,
-                      false,
-                      label.width,
-                      defaultPadding,
-                      null,
-                      null
-                    ) || rendered;
+                  const dimensions = this.calculateImageOrLabelDimensions_(
+                    label.width,
+                    label.height,
+                    part[0],
+                    part[1],
+                    label.width,
+                    label.height,
+                    anchorX,
+                    anchorY,
+                    0,
+                    0,
+                    part[3],
+                    pixelRatioScale,
+                    false,
+                    defaultPadding,
+                    false,
+                    feature
+                  );
+                  if (
+                    opt_declutterTree &&
+                    opt_declutterTree.collides(dimensions.declutterBox)
+                  ) {
+                    break drawChars;
```

Comment:
Another minor issue: The control flow of the function is already quite complicated. Maybe add a label `instructionLoop` to the whole outer while loop and use `++i; continue instructionLoop;` instead? Don't know if that is real improvement.
Then at the other places where continue is used the label can be used as well.

Reply:
Not sure if this would be a real improvement either.